### PR TITLE
Add Metadata support

### PR DESCRIPTION
**Pull request by yuval-k**
*Issue #:* #12

*Description of changes:*
* Adding metadata support in the api & cli
* Added allow mmds to network card (automatically set to enabled in cli if metadata provided)
* Added a test in machine_test.go and tested manually. made sure test pass with `make test`
* Added `make generate` target for convenience and update HACKING.md

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by yuval-k** [comment]
Also note: I updated swagger.yaml due to some differences between it and the implementation.

File: machine.go [hunk 1]
```diff
@@ -530,6 +533,28 @@ func (m *Machine) startInstance(ctx context.Context) error {
 	return err
 }
 
+// SetMetadata sets the machine's metadata for MDDS
+func (m *Machine) SetMetadata(ctx context.Context, metadata interface{}) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
```
Comment (mxpv) [inline machine.go]: Why do we need `cancel` here?

Comment (yuval-k) [reply machine.go]: TBH I copied the pattern from `startInstance`. I suspect it's there to cancel out any kept alive connections, but I don't really know.
I'll be happy to remove this from here and `startInstance` if you think that's the correct thing to do.

File: client/swagger.yaml [hunk 2]
```diff
@@ -393,7 +395,7 @@ definitions:
 
   Error:
     properties:
-      faultMessage:
+      fault_message:
```
Comment (nmeyerhans) [inline client/swagger.yaml]: :100: Thanks for noticing this. Please submit your swagger changes to the firecracker project's [swagger file](https://github.com/firecracker-microvm/firecracker/blob/master/api_server/swagger/firecracker.yaml) as well!

File: cmd/firectl/main.go [hunk 3]
```diff
@@ -131,6 +133,13 @@ func main() {
 		logger.SetLevel(log.DebugLevel)
 	}
 
+	var metadata interface{}
+	if opts.FcMetadata != "" {
+		if err := json.Unmarshal([]byte(opts.FcMetadata), &metadata); err != nil {
+			log.Fatalf("Unable to metadata as json: %s", err)
```
Comment (nmeyerhans) [inline cmd/firectl/main.go]: I think you're missing a word here. "Unable to parse metadata as json", maybe?

File: machine.go [hunk 4]
```diff
@@ -530,6 +533,28 @@ func (m *Machine) startInstance(ctx context.Context) error {
...
+		var message string
+		if respcreated != nil {
+			message = respcreated.Error()
+		}
+		if respnocontent != nil {
+			message = respnocontent.Error()
+		}
+		m.logger.Printf("SetMetadata successful: %s", message)
+	} else {
+		m.logger.Printf("Error setting metadata: %s", err)
```
Comment (nmeyerhans) [inline machine.go]: This should be logged at Error severity, and drop "Error" from the message string.

Comment (yuval-k) [reply machine.go]: this pattern was too copied from `startInstance` - would you like me to change it as well?

File: client/swagger.yaml (hunk 2 above)
Comment (yuval-k) [reply client/swagger.yaml]: https://github.com/firecracker-microvm/firecracker/pull/713

File: machine.go (hunk 1 above)
Comment (mxpv) [reply machine.go]: I don't think it's needed. Please do so.

File: machine.go [hunk 5]
```diff
@@ -518,14 +521,30 @@ func (m *Machine) startInstance(ctx context.Context) error {
...
 
-	ctx, cancel := context.WithCancel(ctx)
-	defer cancel()
-
 	resp, err := m.client.CreateSyncAction(ctx, &info)
 	if err == nil {
 		m.logger.Printf("startInstance successful: %s", resp.Error())
 	} else {
-		m.logger.Printf("Error starting instance: %s", err)
+		m.logger.Errorf("Starting instance: %s", err)
```
Comment (nmeyerhans) [inline machine.go]: :+1: At some point I'll do an audit of all our messages and make sure the severities and language is correct. Thanks for taking care of this one.

**Comment by nmeyerhans** [comment]
I've tested this with `firectl` and it looks good. Let's squash and cleanup the commits before merging, but otherwise, :shipit:!

**Comment by yuval-k** [comment]
usually what we do internall is use github's "Squash and Merge" feature - that way you get a reference to the PR in the commit log

**Comment by yuval-k** [comment]
Alternatively I can squash and force push. let me know what you prefer.

**Comment by nmeyerhans** [comment]
Squash and force-push, please.

**Comment by yuval-k** [comment]
squash everything to one commit and force pushed

**Pull request merged**